### ISSUE TITLE
Managing sensitive environment variables

### ISSUE THREAD

**Issue by ajkitson** [apex/apex#613]
I'm experimenting with moving some existing lambdas to use apex and love it so far. However there's an issue I'm not entirely sure how to handle.

Some of our lambdas have sensitive environment variables--API secrets, DB passwords, etc. We currently use encrypted environment variables to store these. I don't want to put them in the project.json or function.json file.

I know I can use the `--set` and `--env-file` options to set environment variables. But it's tedious to do with each deploy, and if I forget the variables are removed from the new version of the lambda.

Is there some better way to manage sensitive environment variables?

Ideally, I'd be able to specify a file in project.json or function.json that would automatically be used on every deploy but wouldn't be included in the project. Or there would be some way to specify environment variables that should be left in place unless they are overwritten in one of the config files.

**Comment by polds** [comment]
Might not fit your usecase, but I use Consul + Consul Template + Vault for this very purpose. For example here is my "`project.json.ctmpl`":

```
{
  "name": "myproject",
  "description": "Collection of APIs",
  "memory": 128,
  "timeout": 20,
  "role": "arn:aws:iam::123456789012:role/myproject_lambda_function",
  "environment": {
    "HASHIDS_SALT": "somesalt",
  	{{with secret "secret/db" -}}
  	"DB_USER": "{{.Data.user}}",
  	"DB_PASS": "{{.Data.pass}}"
  	{{- end}}
  }
}
```

**Comment by tj** [comment]
I'd be cool with sourcing a file from the project.json. In the meantime maybe:

```
alias apex='apex --env-file ./some.json'
```

**Comment by dblock** [comment]
Maybe there should just be a way for apex to leave the lambda environment variables alone? We typically set server-side variables on, say, Heroku and then just push new code. Is there a philosophical reason why it wouldn't be implemented the same way in apex or at least it would have a flag?

**Comment by tj** [comment]
In Lambda they're defined for each distinct function version. I'm not really against having some sort of way to set them globally but I think it kind of goes against the stateless nature of functions.

**Comment by dblock** [comment]
@tj what's your recommendation with dealing with deployment of some "production" function where we want to set a sensitive key=value?

**Comment by dblock** [comment]
Also note that while environment variables are per version, http://docs.aws.amazon.com/lambda/latest/dg/env_variables.html says that environment variables are managed centrally. If I go and create a new function in the UI I think I'm inheriting these settings, am I not?

**Comment by askot** [comment]
I agree with @dblock, I would like to set/encrypt sensitive variables manually in AWS console and somehow not overwrite them on next deploy, so that I can keep sensitive configuration values out of my CI system.

Maybe an option to not overwrite env variables? Or a list of variable names to merge from existing configuration?

**Comment by tj** [comment]
I'm 99% sure they're coupled to the function's config, to me that makes the most sense really, as the function's requirements change over time as well. In the console you still have to save the function to update the vars etc.

**Comment by mantoni** [comment]
I'm facing the same problem. Currently, I have to set all secret variables on all functions in my deployment script with `--set` or I have to explicitly run apex for each Lambda function with `--set`.

My suggestion would be to replace expressions like `${MY_VAR}` in function config files with the corresponding environment variable. Ideally, the deployment fails if one of the variables can not be resolved.

This would also allow to configure different values per environment. E.g. in `function.prod.json`:

```json
{
  "MY_VAR": "${LAMBDA_MY_VAR_PROD}"
}
```

Does that make sense?

**Comment by natdm** [comment]
Any update on this solution? @mantoni seems to have a decent idea.

**Comment by homanchou** [comment]
How about keeping sensitive credentials in AWS EC2 Systems Manager Parameter Store, that way they are not in source control.  

Then as others have suggested, allow interpolation in the project.json file like:
`{ "FOO": "${ssm.FOO}" }`

When the function is deployed we first make a ssm GetParameters call for all the environment values starting with `${ssm.` and get the real values.  The real values will still be stored with the lambda environment values in plain text, but this is at least one step better than storing the credentials in source control.
